gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java

```diff
@@ -178,31 +178,43 @@ public static void renameRecursively(FileSystem fileSystem, Path from, Path to)
 
       Path toFilePath = new Path(to, relativeFilePath);
 
-      if (!fileSystem.exists(toFilePath)) {
-        /*
-         * Permissions for newly created directories will be inherited from parent directory. This is because the source
-         * and destination directory path may diverge after a certain level
-         *
-         * Source :/data/gobblin/task-staging/writer_1234/Log/file.txt
-         *
-         * Destination :/data/gobblin/task-output/Log/file.txt
-         *
-         * When the directory Log is being renamed from source to destination, task-output directory may not exist. It
-         * will be created by inheriting it's parent permission (/data/gobblin/)
-         */
-        if (!fileSystem.exists(toFilePath.getParent())) {
-          fileSystem.mkdirs(toFilePath.getParent());
-        }
-
-        if (!fileSystem.rename(fromFile.getPath(), toFilePath)) {
-          throw new IOException(String.format("Failed to rename %s to %s.", fromFile.getPath(), toFilePath));
+      if (!safeRenameIfNotExists(fileSystem, fromFile.getPath(), toFilePath)) {
```

Comment:
It's fine, I believe. `safeRenameIfNotExists` returns false if the directory already exists, in which case we will try with the child directory. Assuming that directories are only being created at the target, and not deleted, this should eliminate race conditions.